certbot-nginx/certbot_nginx/configurator.py

```diff
@@ -250,6 +259,121 @@ def _get_ranked_matches(self, target_name):
                                 'rank': 6 if vhost.ssl else 7})
         return sorted(matches, key=lambda x: x['rank'])
 
+    def all_vhost_ports(self, domain):
+        """Get all ports that all vhosts are listening on, that match domain.
+
+        :param str domain: domain to find ports of
+        :returns: set of strings representing the port numbers
+        :rtype: set
+        """
+        matching_vhosts = [item['vhost'] for item in self._get_ranked_matches(domain)]
+        out = set([addr.get_port() for vhost in matching_vhosts for addr in vhost.addrs])
+        # also check the default port, 80
+        out.add(self.DEFAULT_PORT)
+        if "" in out:
+            out.remove("")
+        if r"*" in out:
+            out.remove(r"*")
+        return out
+
+    def choose_redirect_vhost(self, target_name, port):
+        r"""Chooses a virtual host for redirect enhancement.
```

Comment:
Actually, the best way to fix all this `r"\*"` stuff is to take out all code referencing it, because it's wrong! Woohoo!